pkg/volume/fc/fc.go

```diff
@@ -0,0 +1,197 @@
+/*
+Copyright 2015 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package fc
+
+import (
+	"strconv"
+
+	"github.com/golang/glog"
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/types"
+	"k8s.io/kubernetes/pkg/util"
+	"k8s.io/kubernetes/pkg/util/exec"
+	"k8s.io/kubernetes/pkg/util/mount"
+	"k8s.io/kubernetes/pkg/volume"
+)
+
+// This is the primary entrypoint for volume plugins.
+func ProbeVolumePlugins() []volume.VolumePlugin {
+	return []volume.VolumePlugin{&fcPlugin{nil, exec.New()}}
+}
+
+type fcPlugin struct {
+	host volume.VolumeHost
+	exe  exec.Interface
+}
+
+var _ volume.VolumePlugin = &fcPlugin{}
+var _ volume.PersistentVolumePlugin = &fcPlugin{}
+
+const (
+	fcPluginName = "kubernetes.io/fc"
+)
+
+func (plugin *fcPlugin) Init(host volume.VolumeHost) {
+	plugin.host = host
+}
+
+func (plugin *fcPlugin) Name() string {
+	return fcPluginName
+}
+
+func (plugin *fcPlugin) CanSupport(spec *volume.Spec) bool {
+	if spec.VolumeSource.FC == nil && spec.PersistentVolumeSource.FC == nil {
+		return false
+	}
+	// TODO:  turn this into a func so CanSupport can be unit tested without
+	// having to make system calls
```

Comment:
+1 for moving the syscall to a func but ...

I'm thinking of user expectations.  They might have an FC disk (or RBD or expect to use GitRepo, etc) and the binary for the command is not on the machine.

CanSupport returns False in those cases, not error.  When a pod goes to mount the volume, the user gets a "no plugin found" error instead of something specific, such as "FiberChannel not supported on this machine".





I see the same todo in ISCSI.  Missing from RBD.  Same needed for Glusterfs?

Would we need the same for the git volume?  It uses the "git" command that would fail unless installed.